src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_ObjectOrCollectionInitializerExpression.cs

```diff
@@ -442,19 +441,125 @@ private BoundExpression VisitObjectInitializerMember(BoundObjectInitializerMembe
+                    {
+                        rewrittenAccess = VisitIndexerAccess(indexerAccess, isLeftOfAssignment: !isRhsNestedInitializer || indexerAccess.Indexer.RefKind != RefKind.None);
+                    }
+                    else
+                    {
+                        rewrittenAccess = transformedImplicitIndexer;
+                    }
+
+                    if (!isRhsNestedInitializer)
+                    {
+                        var rewrittenRight = VisitExpression(right);
+                        result.Add(MakeStaticAssignmentOperator(assignment.Syntax, rewrittenAccess, rewrittenRight, isRef: false, assignment.Type, used: false));
+                        return;
+                    }
+
+                    break;
+
                 default:
-                    throw ExceptionUtilities.UnexpectedValue((rewrittenLeft ?? assignment.Left).Kind);
+                    throw ExceptionUtilities.UnexpectedValue(left.Kind);
+            }
+
+            AddObjectOrCollectionInitializers(ref dynamicSiteInitializers, ref temps, result, rewrittenAccess, right);
+            return;
+
+            static bool onlyContainsEmptyLeafNestedInitializers(BoundExpression expression)
+            {
+                return expression is BoundObjectInitializerExpression initializer
+                    && initializer.Initializers.All(e => e is BoundAssignmentOperator assignment && onlyContainsEmptyLeafNestedInitializers(assignment.Right));
             }
 
-            AddObjectOrCollectionInitializers(ref dynamicSiteInitializers, ref temps, result, rewrittenAccess, assignment.Right);
+            void addIndexes(ArrayBuilder<BoundExpression> result, BoundAssignmentOperator assignment)
+            {
+                // If we have an element access of the form `[arguments] = { ... }`, we'll evaluate `arguments` only
+                var lhs = assignment.Left;
+                if (lhs is BoundObjectInitializerMember initializerMember)
+                {
+                    foreach (var argument in initializerMember.Arguments)
+                    {
+                        if (argument is BoundArrayCreation { IsParamsArray: true, InitializerOpt: var initializers })
+                        {
+                            Debug.Assert(initializers is not null);
+                            foreach (var element in initializers.Initializers)
+                            {
+                                result.Add(VisitExpression(element));
+                            }
+                        }
+                        else
+                        {
+                            result.Add(VisitExpression(argument));
+                        }
+                    }
+                }
+                else if (lhs is BoundImplicitIndexerAccess implicitIndexerAccess)
+                {
+                    result.Add(VisitExpression(implicitIndexerAccess.Argument));
```

Comment:
>VisitExpression(implicitIndexerAccess.Argument) [](http://example.com/codeflow?start=31&length=47)

Do we want to avoid creating actual ```Index```/```Range``` instances for scenarios for which we do not create them when the initializer is not empty (from end, range expression, etc.)?

Reply:
I'm not planning to do that optimization.